In [50]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Car Rental Dataset Preprocessing

In [51]:
data_path = '../data/raw/car_rental_sample.csv'
raw_df = pd.read_csv(data_path)

## Data Summary

In [52]:
raw_df.head()

airport airport_iata country    city  rental_length  start_date  \
0  Heathrow Airport          LHR      GB  London              2  2023-07-25   
1  Heathrow Airport          LHR      GB  London              2  2023-07-25   
2  Heathrow Airport          LHR      GB  London              2  2023-07-25   
3  Heathrow Airport          LHR      GB  London              2  2023-07-25   
4  Heathrow Airport          LHR      GB  London              2  2023-07-25   

  start_time return_date return_time  date_offset  ...  condition  \
0      10:00  2023-07-27       10:00            0  ...        7.0   
1      10:00  2023-07-27       10:00            0  ...        7.9   
2      10:00  2023-07-27       10:00            0  ...        7.9   
3      10:00  2023-07-27       10:00            0  ...        7.9   
4      10:00  2023-07-27       10:00            0  ...        7.9   

   dropoff_time  efficiency location pickup_time  value_for_money  \
0           5.5         6.3      4.5         4.0              6.1   
1           8.6         6.9      5.8         5.6              6.6   
2           8.6         6.9      5.8         5.6              6.6   
3           8.6         6.9      5.8         5.6              6.6   
4           8.6         6.9      5.8         5.6              6.6   

   no_of_ratings              RunDate  \
0            210  2023-07-23 05:05:17   
1           4909  2023-07-23 05:05:17   
2           4909  2023-07-23 05:05:17   
3           4909  2023-07-23 05:05:17   
4           4909  2023-07-23 05:05:17   

                                         setup_prams      tid  
0  device=Pixel 4&network_type=wifi&languagecode=...  4269268  
1  device=Pixel 4&network_type=wifi&languagecode=...  4269269  
2  device=Pixel 4&network_type=wifi&languagecode=...  4269270  
3  device=Pixel 4&network_type=wifi&languagecode=...  4269271  
4  device=Pixel 4&network_type=wifi&languagecode=...  4269272  

[5 rows x 41 columns]

In [53]:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2730 entries, 0 to 2729
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   airport                2730 non-null   object 
 1   airport_iata           2730 non-null   object 
 2   country                2730 non-null   object 
 3   city                   2730 non-null   object 
 4   rental_length          2730 non-null   int64  
 5   start_date             2730 non-null   object 
 6   start_time             2730 non-null   object 
 7   return_date            2730 non-null   object 
 8   return_time            2730 non-null   object 
 9   date_offset            2730 non-null   int64  
 10  deposit_price          2730 non-null   float64
 11  drive_away_price       2730 non-null   float64
 12  price                  2730 non-null   float64
 13  currency               2730 non-null   object 
 14  product_name           2730 non-null   object 
 15  prod

In [54]:
raw_df.describe(include='all')

airport airport_iata country    city  \
count                                   2730         2730    2730    2730   
unique                                    14           14      11      13   
top     Adolfo Suárez Madrid–Barajas Airport          MAD      ES  Madrid   
freq                                     392          392     734     392   
mean                                     NaN          NaN     NaN     NaN   
std                                      NaN          NaN     NaN     NaN   
min                                      NaN          NaN     NaN     NaN   
25%                                      NaN          NaN     NaN     NaN   
50%                                      NaN          NaN     NaN     NaN   
75%                                      NaN          NaN     NaN     NaN   
max                                      NaN          NaN     NaN     NaN   

        rental_length  start_date start_time return_date return_time  \
count          2730.0        2730       2730        2730        2730   
unique            NaN           1          1           1           1   
top               NaN  2023-07-25      10:00  2023-07-27       10:00   
freq              NaN        2730       2730        2730        2730   
mean              2.0         NaN        NaN         NaN         NaN   
std               0.0         NaN        NaN         NaN         NaN   
min               2.0         NaN        NaN         NaN         NaN   
25%               2.0         NaN        NaN         NaN         NaN   
50%               2.0         NaN        NaN         NaN         NaN   
75%               2.0         NaN        NaN         NaN         NaN   
max               2.0         NaN        NaN         NaN         NaN   

        date_offset  ...    condition  dropoff_time   efficiency     location  \
count        2730.0  ...  2730.000000   2730.000000  2730.000000  2730.000000   
unique          NaN  ...          NaN           NaN          NaN          NaN   
top             NaN  ...          NaN           NaN          NaN          NaN   
freq            NaN  ...          NaN           NaN          NaN          NaN   
mean            0.0  ...     8.240220      8.827326     7.938718     7.761978   
std             0.0  ...     0.989743      1.019272     1.117402     1.164938   
min             0.0  ...     0.000000      0.000000     0.000000     0.000000   
25%             0.0  ...     8.000000      8.700000     7.400000     7.300000   
50%             0.0  ...     8.400000      9.000000     8.100000     8.100000   
75%             0.0  ...     8.700000      9.300000     8.800000     8.400000   
max             0.0  ...    10.000000      9.900000     9.400000     9.300000   

        pickup_time  value_for_money  no_of_ratings              RunDate  \
count   2730.000000       2730.00000    2730.000000                 2730   
unique          NaN              NaN            NaN                    1   
top             NaN              NaN            NaN  2023-07-23 05:05:17   
freq            NaN              NaN            NaN                 2730   
mean       6.897363          7.30978    1949.138095                  NaN   
std        1.157660          0.93887    2017.464056                  NaN   
min        0.000000          0.00000       0.000000                  NaN   
25%        6.500000          7.00000     761.000000                  NaN   
50%        7.100000          7.40000    1423.000000                  NaN   
75%        7.700000          7.90000    2233.000000                  NaN   
max        8.700000          9.80000   14154.000000                  NaN   

                                              setup_prams           tid  
count                                                2730  2.730000e+03  
unique                                                 11           NaN  
top     device=Pixel 4&network_type=wifi&languagecode=...           NaN  
freq                                                  734        

In [55]:
missing_values = raw_df.isnull().sum()
print("Missing or null values in each column:\n", missing_values)

Missing or null values in each column:
 airport                     0
airport_iata                0
country                     0
city                        0
rental_length               0
start_date                  0
start_time                  0
return_date                 0
return_time                 0
date_offset                 0
deposit_price               0
drive_away_price            0
price                       0
currency                    0
product_name                0
product_id                  0
airbags                     0
aircon                      0
free_cancellation           0
doors                       0
group                       0
seats                       0
fuel_type                2730
transmission                0
mileage                     0
supplier_name               0
supplier_address            0
supplier_loction_type       0
average                     0
average_text              103
cleanliness                 0
condition                   0


#### Dropping Missing Values
Since `fuel_type` is the only column with missing values, and all 2730 rows of the data set are missing, we will drop the `fuel_type` column. Missing values for `average_text` can be derived.

In [56]:
cleaned_df = raw_df.drop(columns=['fuel_type'])
missing_values = cleaned_df.isnull().sum()
print("Data after dropping 'fuel_type' column:\n", missing_values)

Data after dropping 'fuel_type' column:
 airport                    0
airport_iata               0
country                    0
city                       0
rental_length              0
start_date                 0
start_time                 0
return_date                0
return_time                0
date_offset                0
deposit_price              0
drive_away_price           0
price                      0
currency                   0
product_name               0
product_id                 0
airbags                    0
aircon                     0
free_cancellation          0
doors                      0
group                      0
seats                      0
transmission               0
mileage                    0
supplier_name              0
supplier_address           0
supplier_loction_type      0
average                    0
average_text             103
cleanliness                0
condition                  0
dropoff_time               0
efficiency                 0
lo

### Non-null columns

In [60]:
missing_values = cleaned_df.isnull().sum()
print("Missing or null values in each column:\n", missing_values)

Missing or null values in each column:
 airport                    0
airport_iata               0
country                    0
city                       0
rental_length              0
start_date                 0
start_time                 0
return_date                0
return_time                0
date_offset                0
deposit_price              0
drive_away_price           0
price                      0
currency                   0
product_name               0
product_id                 0
airbags                    0
aircon                     0
free_cancellation          0
doors                      0
group                      0
seats                      0
transmission               0
mileage                    0
supplier_name              0
supplier_address           0
supplier_loction_type      0
average                    0
average_text             103
cleanliness                0
condition                  0
dropoff_time               0
efficiency                 0
loc

## Data Type Conversions

In [72]:
processed_df = cleaned_df
print("Columns in cleaned_df:", processed_df.columns)


Columns in cleaned_df: Index(['airport', 'airport_iata', 'country', 'city', 'rental_length',
       'start_date', 'start_time', 'return_date', 'return_time', 'date_offset',
       'deposit_price', 'drive_away_price', 'price', 'currency',
       'product_name', 'product_id', 'airbags', 'aircon', 'free_cancellation',
       'doors', 'group', 'seats', 'transmission', 'mileage', 'supplier_name',
       'supplier_address', 'supplier_loction_type', 'average', 'average_text',
       'cleanliness', 'condition', 'dropoff_time', 'efficiency', 'location',
       'pickup_time', 'value_for_money', 'no_of_ratings', 'RunDate',
       'setup_prams', 'tid'],
      dtype='object')


In [ ]:
processed_df['start_date'] = pd.to_datetime(processed_df['start_date'], errors='coerce')
processed_df['return_date'] = pd.to_datetime(processed_df['return_date'], errors='coerce')

print("Number of missing values in 'rental_date':", processed_df['start_date'].isna().sum())
print("Number of missing values in 'return_date':", processed_df['return_date'].isna().sum())

print("Data after converting date columns and handling missing values:\n")
processed_df.head()


Number of NaT values in 'rental_date': 0
Number of NaT values in 'return_date': 0
Data after converting date columns and handling NaT values:



airport airport_iata country    city  rental_length start_date  \
0  Heathrow Airport          LHR      GB  London              2 2023-07-25   
1  Heathrow Airport          LHR      GB  London              2 2023-07-25   
2  Heathrow Airport          LHR      GB  London              2 2023-07-25   
3  Heathrow Airport          LHR      GB  London              2 2023-07-25   
4  Heathrow Airport          LHR      GB  London              2 2023-07-25   

  start_time return_date return_time  date_offset  ...  condition  \
0      10:00  2023-07-27       10:00            0  ...        7.0   
1      10:00  2023-07-27       10:00            0  ...        7.9   
2      10:00  2023-07-27       10:00            0  ...        7.9   
3      10:00  2023-07-27       10:00            0  ...        7.9   
4      10:00  2023-07-27       10:00            0  ...        7.9   

   dropoff_time  efficiency location pickup_time  value_for_money  \
0           5.5         6.3      4.5         4.0              6.1   
1           8.6         6.9      5.8         5.6              6.6   
2           8.6         6.9      5.8         5.6              6.6   
3           8.6         6.9      5.8         5.6              6.6   
4           8.6         6.9      5.8         5.6              6.6   

   no_of_ratings              RunDate  \
0            210  2023-07-23 05:05:17   
1           4909  2023-07-23 05:05:17   
2           4909  2023-07-23 05:05:17   
3           4909  2023-07-23 05:05:17   
4           4909  2023-07-23 05:05:17   

                                         setup_prams      tid  
0  device=Pixel 4&network_type=wifi&languagecode=...  4269268  
1  device=Pixel 4&network_type=wifi&languagecode=...  4269269  
2  device=Pixel 4&network_type=wifi&languagecode=...  4269270  
3  device=Pixel 4&network_type=wifi&languagecode=...  4269271  
4  device=Pixel 4&network_type=wifi&languagecode=...  4269272  

[5 rows x 40 columns]

## Svaing the Cleaned Dataset

In [ ]:
cleaned_file_path = '../data/cleaned/car_rental_sample.csv'
processed_df.to_csv(cleaned_file_path, index=False)

print("\nData Preprocessing Complete. Cleaned data saved to:", cleaned_file_path)


Data Preprocessing Complete. Cleaned data saved to: ../data/cleaned/enterprise_station.csv
